lib/phoenix_live_view/html_engine.ex

```diff
@@ -244,4 +244,307 @@ defmodule Phoenix.LiveView.HTMLEngine do
+    hooks
+  end
+
+  defp rewrite_hook_names(hooks, tokens) do
+    for token <- tokens, reduce: [] do
+      acc ->
+        case token do
+          {:tag, name, attrs, meta} ->
+            [{:tag, name, rewrite_hook_attrs(hooks, attrs), meta} | acc]
+
+          {:local_component, name, attrs, meta} ->
+            [{:local_component, name, rewrite_hook_attrs(hooks, attrs), meta} | acc]
+
+          {:remote_component, name, attrs, meta} ->
+            [{:remote_component, name, rewrite_hook_attrs(hooks, attrs), meta} | acc]
+
+          other ->
+            [other | acc]
+        end
+    end
+  end
+
+  defp rewrite_hook_attrs(hooks, attrs) do
+    Enum.map(attrs, fn
+      {"phx-hook", {:string, name, meta1}, meta2} ->
+        if is_map_key(hooks, name) do
+          {"phx-hook", {:string, hooks[name].name, meta1}, meta2}
+        else
+          {"phx-hook", {:string, name, meta1}, meta2}
+        end
+
+      {attr, value, meta} ->
+        {attr, value, meta}
+    end)
+  end
+
+  defp write_hooks_and_manifest(hooks) do
+    for {_name, %{bundled: true, name: js_filename, content: script_content}} <- hooks do
+      dir = "assets/js/hooks"
+      manifest_path = colocated_hooks_manifest()
+      js_path = Path.join(dir, js_filename <> ".js")
+
+      # colocated hooks are always written to the current otp_app's dir;
+      # so when a dependency is compiled, the hooks will be placed in the deps folder;
+      # but they are still included in the configured manifest, for example:
+      #
+      # /path/to/app/deps/my_dep/assets/js/hooks/HOOKHASH_X_Y.js
+      # /path/to/app/assets/js/hooks/HOOKHASH_Y_Z.js
+      #
+      # the full path is either
+      # ./HOOKHASH_Y_Z.js for a local hook
+      # ../../../deps/my_dep/assets/js/hooks/HOOKHASH_X_Y.js for a hook from a dependency
+      #
+      # the manifest is always ./assets/js/hooks/index.js;
+      # but it can be configured with:
+      #
+      #     config :phoenix_live_view, colocated_hooks_manifest: PATH
+      js_full_path =
+        case Path.relative_to(Path.expand(js_path), Path.dirname(manifest_path), force: true) do
+          <<".", _rest::binary>> = p -> p
+          p -> "./#{p}"
+        end
+
+      File.mkdir_p!(dir)
+      File.write!(js_path, script_content)
+
+      if !File.exists?(manifest_path) do
+        File.write!(manifest_path, """
+        let hooks = {}
+        export default hooks
+        """)
+      end
+
+      manifest = File.read!(manifest_path)
+
+      File.open(manifest_path, [:append], fn file ->
+        if !String.contains?(manifest, js_filename) do
+          IO.puts("Add hook to #{manifest_path}")
+
+          IO.binwrite(
+            file,
+            ~s|\nimport hook_#{js_filename} from "#{js_full_path}"; hooks["#{js_filename}"] = hook_#{js_filename};|
+          )
+        end
+      end)
+
+      IO.puts("Write hook to #{js_path}")
+    end
+  end
+
+  @doc false
+  def prune_hooks(file) do
```

Comment:
I think this has the potential issue in that, if the file is removed altogether, its hooks will still remain.

Reply:
You're right. Do you have another idea?

Reply:
We would need a Mix compiler that checks if modules no longer exist and, if they don't, it cleans up any trailing file.

Reply:
It could be something simple. For example, `assets/hooks/MODULE_NAME`, then we make sure for each MODULE_NAME in hooks there is a beam file.